hdfs/client.py

```diff
@@ -76,22 +84,44 @@ def to_method(self, operation):
 
     def api_handler(client, hdfs_path, data=None, strict=True, **params):
       """Wrapper function."""
-      url = '%s%s%s' % (
-        client.url.rstrip('/'),
-        self.webhdfs_prefix,
-        quote(client.resolve(hdfs_path), '/= '),
-      )
       params['op'] = operation
       if client._proxy is not None:
         params['doas'] = client._proxy
-      return client._request(
-        method=self.method,
-        url=url,
-        data=data,
-        params=params,
-        strict=strict,
-        **self.kwargs
-      )
+
+      attempted_hosts = set()
+      while True:
+        with client._lock:
+          while client._urls[0] in attempted_hosts:
+            client._urls.rotate(-1)
+          host = client._urls[0]
+
+        url = '%s%s%s' % (
+          host.rstrip('/'),
+          self.webhdfs_prefix,
+          quote(client.resolve(hdfs_path), '/= '),
+        )
+
+        try:
+          return client._request(
+            method=self.method,
+            url=url,
+            data=data,
+            params=params,
+            strict=strict,
+            **self.kwargs
+          )
+        except (rq.exceptions.ReadTimeout, rq.exceptions.ConnectTimeout,
+                rq.exceptions.ConnectionError, HdfsError) as exc:
+          if isinstance(exc, HdfsError) and exc.exception != 'StandbyException':
+            raise exc
+          else:
```

Comment:
No need to nest the rest of this block inside an `else`, it'll keep indentation more manageable.